test-data/unit/check-protocols.test

```diff
@@ -0,0 +1,1827 @@
+
+@runtime
+class P(Protocol):
+    def meth(self) -> int:
+        pass
+
+x: object
+if isinstance(x, P):
+    reveal_type(x)  # E: Revealed type is '__main__.P'
+    reveal_type(x.meth())  # E: Revealed type is 'builtins.int'
+
+class C:
+    def meth(self) -> int:
+        pass
+
+z: P = C()
+[builtins fixtures/dict.pyi]
+
+[case testProtocolsCannotInheritFromNormal]
+from typing import Protocol
+
+class C: pass
+class D: pass
+
+class P(C, Protocol): # E: All bases of a protocol must be protocols
+    attr: int
+
+class P2(P, D, Protocol): # E: All bases of a protocol must be protocols
+    pass
+
+P2() # E: Cannot instantiate abstract class 'P2' with abstract attribute 'attr'
+p: P2
+reveal_type(p.attr) # E: Revealed type is 'builtins.int'
+
+-- Generic protocol types
+-- ----------------------
+
+[case testAutomaticProtocolVariance]
+from typing import TypeVar, Protocol
+
+T = TypeVar('T')
+
+# In case of these errors we proceed with declared variance.
+class Pco(Protocol[T]): # E: Invariant type variable 'T' used in protocol where covariant one is expected
+    def meth(self) -> T:
+        pass
+class Pcontra(Protocol[T]): # E: Invariant type variable 'T' used in protocol where contravariant one is expected
+    def meth(self, x: T) -> None:
+        pass
+class Pinv(Protocol[T]):
+    attr: T
+
+class A: pass
+class B(A): pass
+
+x1: Pco[B]
+y1: Pco[A]
+x1 = y1 # E: Incompatible types in assignment (expression has type Pco[A], variable has type Pco[B])
+y1 = x1 # E: Incompatible types in assignment (expression has type Pco[B], variable has type Pco[A])
+
+x2: Pcontra[B]
+y2: Pcontra[A]
+y2 = x2 # E: Incompatible types in assignment (expression has type Pcontra[B], variable has type Pcontra[A])
+x2 = y2 # E: Incompatible types in assignment (expression has type Pcontra[A], variable has type Pcontra[B])
+
+x3: Pinv[B]
+y3: Pinv[A]
+y3 = x3 # E: Incompatible types in assignment (expression has type Pinv[B], variable has type Pinv[A])
+x3 = y3 # E: Incompatible types in assignment (expression has type Pinv[A], variable has type Pinv[B])
+
+[case testProtocolVarianceWithCallableAndList]
+from typing import Protocol, TypeVar, Callable, List
+T = TypeVar('T')
+S = TypeVar('S')
+T_co = TypeVar('T_co', covariant=True)
+
+class P(Protocol[T, S]): # E: Invariant type variable 'T' used in protocol where covariant one is expected \
+                         # E: Invariant type variable 'S' used in protocol where contravariant one is expected
+    def fun(self, callback: Callable[[T], S]) -> None: pass
+
+class P2(Protocol[T_co]): # E: Covariant type variable 'T_co' used in protocol where invariant one is expected
+    lst: List[T_co]
+[builtins fixtures/list.pyi]
+
+[case testProtocolVarianceWithUnusedVariable]
+from typing import Protocol, TypeVar
+T = TypeVar('T')
+
+class P(Protocol[T]): # E: Invariant type variable 'T' used in protocol where covariant one is expected
+    attr: int
+
+[case testGenericProtocolsInference1]
```

Comment:
I don't remember if I saw a test case that would have a related case where the protocol is non-generic and a non-generic class has a generic method. Say, protocol has signature `int -> int` and class has signature `T -> T` (but `T` has method scope).